core/arch/arm/kernel/elf_load.c

```diff
@@ -0,0 +1,371 @@
+	 * the TA context to enable mapping the TA. Later when this
+	 * function has returned and the hash has been verified the flags
+	 * field will be updated with eventual other flags.
+	 */
+	if (state->phdr[0].p_filesz < head_size)
+		return TEE_ERROR_BAD_FORMAT;
+	res = alloc_and_copy_to(&p, state, state->phdr[0].p_offset, head_size);
+	if (res != TEE_SUCCESS)
+		return res;
+	state->ta_head = p;
+	state->ta_head_size = head_size;
+
+	*head = state->ta_head;
+	*vasize = state->vasize;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result el_process_rel(struct elf_load_state *state, size_t rel_sidx,
+			vaddr_t vabase)
+{
+	Elf32_Rel *rel;
+	Elf32_Rel *rel_end;
+
+	if (state->shdr[rel_sidx].sh_entsize != sizeof(Elf32_Rel))
+		return TEE_ERROR_BAD_FORMAT;
+
+	rel = (Elf32_Rel *)(vabase + state->shdr[rel_sidx].sh_addr);
+	if (!TEE_ALIGNMENT_IS_OK(rel, Elf32_Rel))
+		return TEE_ERROR_BAD_FORMAT;
+
+	rel_end = rel + state->shdr[rel_sidx].sh_size / sizeof(Elf32_Rel);
+	for (; rel < rel_end; rel++) {
+		Elf32_Addr *where = (Elf32_Addr *)(vabase + rel->r_offset);
+		size_t sym_idx;
+
+		if (!TEE_ALIGNMENT_IS_OK(where, Elf32_Addr))
+			return TEE_ERROR_BAD_FORMAT;
+
+		switch (ELF32_R_TYPE(rel->r_info)) {
+		case R_ARM_ABS32:
+			sym_idx = ELF32_R_SYM(rel->r_info);
+			if (sym_idx >= state->sym_tab_num_syms)
+				return TEE_ERROR_BAD_FORMAT;
+
+			*where += vabase + state->sym_tab[sym_idx].st_value;
+			break;
+		case R_ARM_RELATIVE:
+			*where += vabase;
+			break;
+		default:
+			EMSG("Unknown relocation type %d",
+			     ELF32_R_TYPE(rel->r_info));
+			return TEE_ERROR_BAD_FORMAT;
+		}
+	}
+	return TEE_SUCCESS;
+}
+
+TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
+{
+	TEE_Result res;
+	size_t n;
+	void *p;
+	uint8_t *dst = (uint8_t *)vabase;
+	Elf32_Ehdr *ehdr = state->ehdr;
+	Elf32_Phdr *phdr = state->phdr;
+
+	/*
+	 * Zero initialize everything to make sure that all memory not
+	 * updated from the ELF is zero (covering .bss and eventual gaps).
+	 */
+	memset(dst, 0, state->vasize);
+
+	/*
+	 * Copy the segments
+	 */
+	memcpy(dst, state->ta_head, state->ta_head_size);
+	res = copy_to(state, dst, state->vasize,
+		      phdr[0].p_vaddr + state->ta_head_size,
+		      phdr[0].p_offset + state->ta_head_size,
+		      phdr[0].p_filesz - state->ta_head_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	for (n = 1; n < ehdr->e_phnum; n++) {
+		if (phdr[n].p_type != PT_LOAD)
+			continue;
+
+		res = copy_to(state, dst, state->vasize, phdr[n].p_vaddr,
+			      phdr[n].p_offset, phdr[n].p_filesz);
+		if (res != TEE_SUCCESS)
+			return res;
+	}
+
+	/*
+	 * We have now loaded all segments into TA memory, now we need to
+	 * process relocation information. To find relocation information
+	 * we need to locate the section headers. The section headers are
+	 * located somewhere between the last segment and the end of the
+	 * ELF.
+	 */
+	if (ehdr->e_shoff < phdr[1].p_filesz)
```

Comment:
What if the program header table has only 1 entry (`ehdr->e_phnum == 1)`? `phdr[1]` would not exist. Anyway, is this test really needed?